flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java

```diff
@@ -907,20 +909,37 @@ public void runIfState(State expectedState, Runnable action, Duration delay) {
 
     // ----------------------------------------------------------------
 
+    /** Note: Do not call this method from a State constructor. */
     @VisibleForTesting
-    void transitionToState(State newState) {
-        if (state != newState) {
-            LOG.debug(
-                    "Transition from state {} to {}.",
-                    state.getClass().getSimpleName(),
-                    newState.getClass().getSimpleName());
-
-            State oldState = state;
-            oldState.onLeave(newState.getClass());
-
-            state = newState;
-            newState.onEnter();
-        }
+    <S extends State> void transitionToState(StateFactory<S> targetState) {
+        Preconditions.checkState(
+                state != null, "State transitions are now allowed while construcing a state.");
+        Preconditions.checkState(
+                state.getClass() != targetState.getStateClass(),
+                "Attempted to transition into the very state the scheduler is already in.");
+
+        LOG.debug(
+                "Transition from state {} to {}.",
+                state.getClass().getSimpleName(),
+                targetState.getStateClass().getSimpleName());
+
+        State oldState = state;
+        oldState.onLeave(targetState.getStateClass());
+
+        // Guard against state transitions while constructing state objects.
+        //
+        // Consider the following scenario:
+        // Scheduler is in state Restarting, once the cancellation is complete, we enter the
+        // transitionToState(WaitingForResources) method.
+        // In the constructor of WaitingForResources, we call `notifyNewResourcesAvailable()`, which
+        // finds resources and enters transitionsToState(Executing). We are in state Executing. Then
+        // we return from the methods and go back in our call stack to the
+        // transitionToState(WaitingForResources) call, where we overwrite Executing with
+        // WaitingForResources. And there we have it, a deployed execution graph, and a scheduler
+        // that is in WaitingForResources.
+        state = null;
```

Comment:
hmm well this is annoying; these kind of loops are driving me crazy in the slot procotol...

what if all goToX methods would schedule the state transition instead, or if transitionToState just schedules it? well that would be annoying during tests wouldn't it...

Reply:
It would probably work with the tests, since we collect and manually trigger all scheduled executions in the close() method, before checking if the expected state transitions happened.

But I'm not sure if it's worth introducing such behavior for the sake of removing the onEnter() method.
One of the reasons why we agreed to remove onEnter() was because the change revealed a few test inaccuracies. Basically all of these inaccuracies have been found and fixed during the review of the respective tests. The onEnter() prototype was just done on an unreviewed version of the tests.

Right now, WaitingForResources is the only state that triggers state transitions in the constructor.

I would actually propose to revisit the decision to remove the onEnter() method, given the compromises we have to do to make it work (CC @tillrohrmann )

Reply:
Theoretically, we could remember in the `transitionToState` method whether we are currently doing a state transition. If this is the case, then we enqueue the `targetState` in some queue. If not, then we set the flag that we are transitioning into a new state and do the state transition. Once we are complete with a state transition we check the queue for enqueued state transitions. Only if all state transitions are dequeued, we switch the flag back to not doing a state transition.

Reply:
"The issues are fixed now; we don't need safeguards anymore" could just as well be used as an argument to keep the PR as is and even remove state transitions check. We fixed the one problematic case and could call it a day.

Overall, my impression is that we should not allow immediate state transitions in the constructor, `onEnter`, or `onLeave`, in any case. Because all of these result in weird loops/interleaving of state transitions that can lead to subtle issues.
IOW, `transitionToState` should be an atomic operation that fully completes before another transition can be triggered.

Hence, whether `onEnter` exists or not is actually not relevant in this consideration. While it was the sole case where this issue occurred, the underlying issues are unclear and unenforced contracts as to what a State is allowed to do in which methods.

As you have shown in the PR it is pretty easy to safeguard against such occurrences; you'd just need to null the state before calling onLeave.

Alternatively, this would also work, and would be a bit more consolidated:
```
oldState = state
state = null;
oldState.onLeave()
newState = targetState.getState()
newState.onEnter()
Preconditions.checkState(state == null); // this will fail if any other state transition occurred in the mean time
state = newState;
```

And as far as I'm concerned that's a pretty tiny cost compared to the risk of testing entirely theoretical scenarios or breaking the state machine.

Reply:
Maybe we could even make it a bit more explicit by having a dedicated flag which we check before doing the state transition. That way the program would directly fail when a new state transition is triggered and not later when the outer state transition will be continued.

Reply:
Wouldn't we achieve this "fail fast" behavior if we'd move the Precondition up in Chesnay's proposal?

```
Preconditions.checkState(state == null); // this will fail if a new state transition is triggered
oldState = state
state = null;
oldState.onLeave()
newState = targetState.getState()
newState.onEnter()
state = newState;
```

Reply:
yes that I think that would also work.

Reply:
maybe a flag is easier to read though
```
this.isTransitioningState = true
this.state.onLeave()
this.state = targetState.getState()
this.state = onEnter()
this.isTransitioningState = false; // maybe this should be done in a finally block...what actually happens if a transition throws an exception?
```